Update several things in the formatting of the documentation #158

santiagogaray opened the pull request:
@chriswmackey @mostaphaRoudsari @theo-armour 
Took a while but I finally completed a first pass on the entire package. I fixed many display/formatting issues but a few things remain. If you have time read on, help is appreciated.

1a)
to_dict() method:
Throughout the package, in many from_dict methods, the docstring shows the dictionary with the description of the keys, this makes sphinx to invalidate it as  json code. 
Could possibly solve it in a number of different ways, just a couple :
a) Not using json block and use a literal or shell block.
b) As shown in ladybug.analysisperiod.AnalysisPeriod: writing the dictionary with default values only and listing the keys outside of the dictionary.

1b) 
Many from_dict() cases fail to be detected as json given the differences between dictionaries and json formats:
Python uses single and double quotes for string literals. JSON only allows double quotes.
Tuples not supported in JSON
Python uses None objects, JSON uses null to as a special "doesn't exist" sentinel value.
Python uses True and False as boolean values, JSON uses true and false.
Python dictionary keys can be any hashable type, JSON only supports strings.
Do we JSONize the python dictionaries or we specify the block as something else(ex. code).

2)
In many classes I added properties when missing. In some classes  Properties / Arguments that listed _init_ inputs were renamed as Args before adding the properties. 
Do we add descriptions to the properties?  We have inconsistency at this point, sometimes in Arguments we have descriptions but not in properties (sometimes are the same, more often than not they differ)

3)
Usage - Is correct to use a literal block for these? Should title be bolded?

4)
Are constants displayed correctly? Right now contents are shown bolded, in some modules they extend multiple lines (ex. ladybug.epw.EPW.FIELDS)


6) 
In Class properties where descriptions extend multiple lines (ex. ladybug.analysisperiod.AnalysisPeriod)  lines 2 and beyond display incorrectly once the property name is made bolded. We actually have this issue in one instance in ladybug.geometry(geometry3d.arc for reference). After trying to solve this in the docstring itself, the only way I seem to find a workaround was by joining these multi-line descriptions in one single line. This was done at the  autodoc_process_docstring by adding a new function, make_properties_one_liner, to execute before make_fields_bolded. I may find a way to simplify this routine but seems to be doing the job well at this point. 
If you think there is a simpler way to tackle this issue feel free to pitch in, any comments is appreciated.
 
7) 
Minor observations - not documentation related:
ladybug.datacollection.HourlyContinuousCollection: Has is_continuous and isContinuous properties

VolumeFlowRate:
Shouldn't isFlowRate property be isVolumeFlowRate?

santiagogaray commented:
from_dict()  possible solutions

![image](https://user-images.githubusercontent.com/19495783/66401609-6be28000-e9b1-11e9-8b11-49e74d9606ea.png)
![image](https://user-images.githubusercontent.com/19495783/66401885-e6ab9b00-e9b1-11e9-898c-5136617df19b.png)

santiagogaray commented:
Usage display

![image](https://user-images.githubusercontent.com/19495783/66402048-2b373680-e9b2-11e9-9818-405bda6a0273.png)

santiagogaray commented:
Constants display

![image](https://user-images.githubusercontent.com/19495783/66402209-7e10ee00-e9b2-11e9-8f7f-a1b83cf68cc2.png)

santiagogaray commented:
Multiline property issue. We have this one case in ladybug-geometry. 

![image](https://user-images.githubusercontent.com/19495783/66402634-2b840180-e9b3-11e9-9e58-053243e75b47.png)

Fixed in this package

chriswmackey commented:
Thank you very much for getting this together @santiagogaray . I will try to review it as soon as I can.

chriswmackey commented:
Thank you again for getting this together @santiagogaray . I am going to answer your high-level questions first and then do a review looking through all of your changes.

1. I think we want to standardize the display of dictionaries in docstrings and I think the convention we used in `ladybug-geometry` is better. So, if you can add a sample value for each key in the docstring dictionary (like a float, int, empty list, empty dictionary) and then put the description behind a comment like `\\` or `#`, that would be ideal.  BTW, I am thinking that we should probably be using `codeblock::python` instead of `codeblock::json`. The input dictionary for this method is a Python one after all, even though it is easily convert-able to a JSON format. So using `codeblock::python` will solve all of the issues you experience there about the difference between Python and JSON.

2. This repo is the oldest of all the Ladybug Tools organization repos and so it has a lot of docstring stuff that is not following the standards we are trying to follow today.  The `ladybug-geometry` repo is much newer and so I would generally say that, when in doubt, do what we did on that repo. For clarification, this means: 1. Adding a bulleted list of properties (without descriptions) at the top of each class. 2. Using an `Args:` tag at the top of the class that describes each input going into the `__init__` method. 3. Including the descriptions of each property on the docstring of that property (not within the bulleted list at the top).

3. Usage should always be in `codeblock::python`.

4. The `FIELDS` constant really is not that helpful and I think we should make it hidden because, otherwise, it makes the docs really unreadable. So changing `FIELDS` to `_fields` should do the trick in this situation.  For the other constants, I think we usually want them to show up in the docs (like I know I definitely want to see the 'VALIDTIMESTEPS' constant on the `AnalysisPeriod` to show up). For this situation, I think the boded text makes sense. Feel free to tweak it if you think it can be improved.

6. I think this one should be solved by following my suggestion above. Namely, let's refrain from having a description of the property within the bulleted list and instead put the description of the property within the docstring of the property itself.

7. Thank you for the observations. Actually, all of these situations were we have `isSuchAndSuch` properties is not good practice and we should remove them at some point soon. It's a part of a hack that Mostapha had done to get the library to work in the (now already broken) Dynamo version, which we realize now is not worth making the code here un-clean. We will have to remove both these `isSuchAndSuch` properties and any cases where they are used in `hasattr` methods throughout the library. We can do this all in a later PR, though, so you don't have to worry about it now. Perhaps the only bearing that this has on your docstrings is that you can leave these `isSuchAndSuch` properties out of the bulleted list of properties at the top.

chriswmackey commented on docs/conf.py:
```diff
@@ -213,3 +215,119 @@
+    return lines
+
+
+def make_properties_one_liner(lines):
```
As I said in my previous answers to your questions, I think we can remove this and just use better practices of not describing inputs in the bulleted list of properties.

santiagogaray commented:
Thanks @chriswmackey for your comments. I have updated the package with the changes suggested. 

A couple of things to check:

-The values used in the dictionaries can be example values, empty data structures, or variables. Feel free to indicate if any of these approaches are not working.

- I removed all the property descriptions included in the DataTypeBase class. Given these descriptions are more extensive than the ones included in the property methods, I used the former descriptions to replace the existing ones. Is this amount of information correct to be included here?

- As you can see the conf.py got a lot simpler now, without any property descriptions there is no need for any post-processing tasks . I'm assuming this will have to be updated in the ladybug_geometry package to be consistent, right?

Thanks again Chris, look forward to your comments.

chriswmackey commented:
@santiagogaray ,

Thanks for addressing the comments and I think we still have one more round of me answering high-level questions before I dig in deep.

Firs and foremost, you will see that one of your recent changes has broken the code, causing a particularly high number of failures in the `EPW` and `Wea` classes that you can see here: https://travis-ci.org/ladybug-tools/ladybug/jobs/596200537
I think this might have resulted from you making changes from `FIELDS` to `_fields` without changing all of the places where the original `FIELDS` constant is referenced.  We will need to get this fixed before I can do a deep review and you can run the tests locally from command line by using the following steps:
1. Install pytest by typing `pip install pytest`
2. Navigate your command window to the root of the reop on your local machine.
3. Type `pytest .\tests`
Then you should see all the tests will run and you'll get info about all of the errors. Note that you may need to include `python -m` before each of your commands depending on how Python has been added to your PATH.

I'll answer your other questions shortly.

santiagogaray commented:
Thanks @chriswmackey for guiding me with the pytest tool. After renaming a couple of FIELDS I checked again and no errors reported.  Happy to learn yet a new thing related to this process!

1. I did not explain myself well in regards to the dictionaries and sample values.  I will show with examples in a separate comment below.
 
2. I did tweak a few properties' docstrings in DataTypeBase. Hope is better, happy to simplify more if needed.

3. I had fun learning regex, thanks for giving the opportunity to learn it!

santiagogaray commented:
When deciding the dictionary sample values I was using variable names instead of actual values in dictionaries where some keys hold some unknown data structure (ex. header below). What would you do in this case? Is header here an empty dictionary?

![image](https://user-images.githubusercontent.com/19495783/66611470-ace2bc00-eb8c-11e9-9535-94054b5a0320.png)

chriswmackey commented:
Thank you for fixing the failing tests @santiagogaray . As you can see, the automated run of the tests with each new commit is important and I'm happy to show you more about how they work at a later point.

Ah, that's what you meant by the dictionaries. For these, what you have there is close but, for the case where you have another object nested within the parent one, the dictionary representation will have another nested dictionary. So instead of `h`, you should use some empty curly brackets `{}` to denote that the value there is another dictionary. Similarly, in the case that the object is an array of something, you should use empty square brackets `[]`.  Altogether, the dictionary that you screen-shot there should look like this:
```
                {
                "header": {}  # A Ladybug Header,
                "values": []  # An array of values,
                "datetimes": []  # An array of datetimes,
                "validated_a_period": True  # Boolean for whether analysis_period is valid
                }
```

BTW, I forgot to confirm with you that it will be good to go back to ladybug-geometry and remove the descriptions from the bulleted list of properties. Feel free to send a PR for that whenever you want and we should be able to merge it in quickly.

chriswmackey commented:
Also, once you fix those dictionary displays, I'll do a final review.

santiagogaray commented:
Thanks @chriswmackey for clarifying. I hope I got this right this time and you can start your review.

I will get going on the ladybug_geoemtry package soon.

I appreciate the patience and time to guide me through all the steps involved in this. I'm grateful I have a chance to learn as we go through things. Thanks!

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1136,11 +1205,21 @@ def __repr__(self):
         day_of_month
         daylight_savings_indicator
+        beam_shced
+        diff_sched
```
I also forgot the docstring here. Can you add the following:
```
        solar_model: Text for the name of the solar model to use. Choose from the
            following: 'ASHRAEClearSky', 'ASHRAETau', 'ZhangHuang', 'Schedule'
        month: Month in which the design day occurs.
        day_of_month: Day of the month on which the design day occurs.
        daylight_savings_indicator: Text ('Yes' or 'No'), for whether daylight savings
            time is active. Default: 'No'
        beam_shced: Schedule name for beam irradiance. Shoulb be an empty string unless
            solar_model is 'Schedule'.
        diff_sched: Schedule name for diffuse irradiance. Shoulb be an empty string
            unless solar_model is 'Schedule'.
```

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1012,11 +1066,21 @@ def __repr__(self):
         wind_direction
         rain
         snow_on_ground
+
```
I forgot the docstring here. Can you add the following:
```
wind_speed: Wind speed on the design day [m/s].
        wind_direction: Wind direction on the design day [degrees]. Default: 0
        rain: Boolean to indicate rain on the design day. Default: False.
        snow_on_ground: Boolean to indicate snow on the design day. Default: False.
```

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1273,11 +1357,17 @@ def __repr__(self):
         day_of_month
         clearness
         daylight_savings_indicator
+
```
I see I forgot the docstring here. Can you add the following:
```
        month: Month in which the design day occurs.
        day_of_month: Day of the month on which the design day occurs.
        clearness: Value between 0 and 1.2 that will get multiplied by the model's
            irradinace to correct for factors like elevation.
        daylight_savings_indicator: Text ('Yes' or 'No'), for whether daylight savings
            time is active. Default: 'No'
```

santiagogaray replied:
Done!

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1369,12 +1464,22 @@ def to_dict(self):
         tau_b
         tau_d
         daylight_savings_indicator
+
```
I see I forgot a docstring here. Can you add the following:
```
        month: Month in which the design day occurs.
        day_of_month: Day of the month on which the design day occurs.
        tau_b: Value for the 'beam' term in the Tau model. Typically
            found in .stat files.
        tau_d: Value for the 'diffuse' term in the Tau model. Typically
            found in .stat files.
        daylight_savings_indicator: Text ('Yes' or 'No'), for whether daylight savings
            time is active. Default: 'No'
```

santiagogaray replied:
Done!

chriswmackey commented on ladybug/dt.py:
```diff
@@ -256,23 +271,26 @@ def __repr__(self):
+        month: A value for month between 1-12 (Defualt: 1).
+        day: A value for day between 1-31 (Defualt: 1).
+        leap_year: A boolean to indicate if date is for a leap year
+            (Default: False).
```
Let's remove the parentheses around the Default and just add a period after `year`. This is the new standard we are trying to follow throughout the libraries.

chriswmackey commented on ladybug/stat.py:
```diff
@@ -120,33 +123,37 @@ def __init__(self, file_path):
+                {
+                "location": {},  # ladybug location schema
+                "ashrae_climate_zone": "",  # str
+                "koppen_climate_zone": "", # str
```
We can put some better defaults here. How about `"5A"` for `ashrae_climate_zone` and `"Dfa"` for `koppen_climate_zone`. Those are the values for Boston.

chriswmackey commented on ladybug/stat.py:
```diff
@@ -120,33 +123,37 @@ def __init__(self, file_path):
+                "monthly_wind": [],  # list of 12 float values for each month
+                "monthly_wind_dirs": [],  # matrix with 12 cols for months of the year
+                                          #and 8 rows for the cardinal directions.
+                "standard_pressure_at_elev": 0.0,  # float value for pressure in Pa
```
A better default here is `101325`, which is air pressure at sea level in pascals. BTW, generally excellent job formatting these dictionaries.

santiagogaray commented on ladybug/datacollection.py:
```diff
@@ -2,23 +2,24 @@
+* max, min, bounds, average, median, total, get_percentile, get_highest/lowest values
+* perform unit conversions on the data: to_unit, to_ip, to_si
+* filter based on conditional statements
+* filter based on analysis period
```
Just needed to be tagged to right. Done!

santiagogaray commented:
Thanks @chriswmackey for putting time to make this detailed review. I hope I fixed all the comments, but please feel free to check if anything else needs to be done. 

I agree the documentation looks great. Seeing Sphinx and bootstrap working together by capturing the docstrings is really special. I'm very excited that I've been able to contribute and learn though this project. 

Tomorrow I will go back to the ladybug-package and make those adjustments  as we discussed!

chriswmackey commented:
Great! Thank you @santiagogaray ! I will do a final build of the docs once I get home and, if all looks good, we'll merge it in.

chriswmackey commented on ladybug/designday.py:
```diff
@@ -31,9 +31,27 @@
+        * location
+        * name
+        * sky_condition
+        * wind_condition
```
I just realized that these properties are in the incorrect location. They shouldn't be non the `DDY` class and they should instead be on the `DesignDay` class.  The only properties that should appear here on the `DDY` object are:
```
* file_path
* location
* design_days
```

chriswmackey commented on ladybug/designday.py:
```diff
@@ -222,6 +245,15 @@ class DesignDay(object):
+        * dry_bulb_condition
+        * humidity_condition
+        * wind_condition
+        * sky_condition
```
As noted above, this list should be extended with the properties currently listed in the `DDY` class docstring.

chriswmackey commented on ladybug/designday.py:
```diff
@@ -866,13 +911,23 @@ def __repr__(self):
+        * wet_bulb_range
+        * hourly_pressure
+        * humid_types
+        * isHumidityCondition
```
Remove `isHumidityCondition` from this list since the whole property is going to be removed soon.

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1012,11 +1072,21 @@ def __repr__(self):
+        * snow_on_ground
+        * hourly_values
+        * hourly_wind_dirs
+        * isWindCondition
```
Remove `isWindCondition`

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1136,11 +1211,25 @@ def __repr__(self):
+        * month
+        * day_of_month
+        * hourly_sky_cover
+        * isSkyCondition
```
Remove `isSkyCondition`

chriswmackey commented on ladybug/designday.py:
```diff
@@ -1369,12 +1476,25 @@ def to_dict(self):
-        daylight_savings_indicator
+        * day_of_month
+        * hourly_sky_cover
+        * isSkyCondition
```
Remove `isSkyCondition`

chriswmackey commented on ladybug/dt.py:
```diff
@@ -8,31 +8,39 @@
+        * hoy
+        * int_hoy
+        * max
+        * microsecond
```
I know that, because the ladybug DateTime inherits from the native Python datetime , this `microsecond` property exists on the class. However, it is not one that is actually usable in any Ladybug workflows and, if users set this to be something other than 0, it can actually mess up certain workflows. So we would prefer to not list it here in the properties, even though I know we can't easily stop it from appearing in the docs. The same thing can be about the following properties, which should be removed from this list (as well as the `Date` and `Time` property lists):
```
* max
* min
* second
* resolution
```
Everything else can stay.

santiagogaray replied:
Done!

chriswmackey commented on ladybug/header.py:
```diff
@@ -14,26 +14,26 @@ class Header(object):
+        * unit
+        * analysis_period
+        * metadata
+        * isHeader
```
Remove `isHeader`

chriswmackey commented on ladybug/datacollection.py:
```diff
@@ -1280,20 +1380,38 @@ def __repr__(self):
+        * header
+        * isDataCollection
+        * isMonthlyPerHour
+        * isHourly
```
Remove the following:
```
        * isDataCollection
        * isMonthlyPerHour
        * isHourly
```

chriswmackey commented on ladybug/datacollection.py:
```diff
@@ -1148,20 +1227,38 @@ def __repr__(self):
+        * header
+        * isDataCollection
+        * isMonthlyPerHour
+        * isHourly
```
Remove the following:
```
        * isDataCollection
        * isMonthlyPerHour
        * isHourly
```

chriswmackey commented on ladybug/datacollection.py:
```diff
@@ -935,20 +993,37 @@ def __repr__(self):
+        * header
+        * isDataCollection
+        * isMonthlyPerHour
+        * isHourly
```
Remove the following:
```
        * isDataCollection
        * isMonthlyPerHour
        * isHourly
```

chriswmackey commented on ladybug/datacollection.py:
```diff
@@ -509,19 +543,39 @@ def __repr__(self):
+        * header
+        * isContinuous
+        * isDataCollection
+        * isHourly
```
Remove the following:
```
        * isDataCollection
        * isMonthlyPerHour
        * isHourly
```

chriswmackey commented on ladybug/datacollection.py:
```diff
@@ -47,19 +48,39 @@
+        * datetimes
+        * header
+        * isDataCollection
+        * isHourly
```
Remove the following:
```
        * isDataCollection
        * isHourly
```

chriswmackey commented on ladybug/designday.py:
```diff
@@ -31,9 +31,27 @@
+        * hourly_wind_direction
+        * hourly_wind_speed
+        * humidity_condition
+        * isDesignDay
```
Remove `isDesignDay`

chriswmackey commented on ladybug/designday.py:
```diff
@@ -764,11 +791,24 @@ def __repr__(self):
+        * dry_bulb_max
+        * dry_bulb_range
+        * hourly_values
+        * isDryBulbCondition
```
Remove `isDryBulbCondition`

chriswmackey commented on ladybug/location.py:
```diff
@@ -18,6 +18,20 @@ class Location(object):
+        * country
+        * elevation
+        * ep_style_location_string
+        * isLocation
```
Remove `isLocation`

chriswmackey commented on ladybug/stat.py:
```diff
@@ -23,36 +23,42 @@
+        * monthly_clear_sky_conditions
+        * monthly_tau_beam
+        * monthly_tau_diffuse
+        * isStat
```
Remove `isStat`

chriswmackey commented on ladybug/wea.py:
```diff
@@ -49,6 +49,18 @@ class Wea(object):
+        * global_horizontal_irradiance
+        * header
+        * hoys
+        * isWea
```
Remove `isWea`

santiagogaray commented:
@chriswmackey 

Thanks for the detailed review Chris. I think I addressed all the edits but let me know if I missed something or if you have additional comments!

chriswmackey commented:
This is wonderful @santiagogaray . Thank you, again for doing this. I'll talk to Mostapha later this week to discuss the next best steps for updating documentation and, soon after I pass a big deadline on Nov 8, I will add some new modules to this repo here that show fully how we plan to make cross-platform climate graphics (opening the door for the other project we talked about).

Merged!

santiagogaray commented:
Sounds good @chriswmackey. I'm very happy I've been able to help in the documentation of these packages and I really look forward to keep helping on these while it makes sense to you guys.  Also happy to get into the big project when things are ready for it!
